operations/host_spawn.go

```diff
@@ -792,3 +796,367 @@ func hostRunCommand() cli.Command {
+	if err != nil {
+		return "", "", errors.Wrap(err, "problem getting your spawn hosts")
+	}
+
+	for _, h := range hosts {
+		if model.FromStringPtr(h.Id) == hostID {
+			catcher := grip.NewBasicCatcher()
+			user = model.FromStringPtr(h.User)
+			catcher.ErrorfWhen(user == "", "could not find login user for host '%s'", hostID)
+			hostname = model.FromStringPtr(h.HostURL)
+			catcher.ErrorfWhen(hostname == "", "could not find hostname for host '%s'", hostID)
+			return user, hostname, catcher.Resolve()
+		}
+	}
+	return "", "", errors.Errorf("could not find host '%s' in user's spawn hosts", hostID)
+}
+
+// sanityCheckRsync performs some basic sanity checks for the common case in
+// which you want to mirror two directories. The trailing slash means to
+// overwrite all of the contents of the destination directory, so we check that
+// they really want to do this.
+func sanityCheckRsync(localPath, remotePath string, pull bool) (ok bool, err error) {
+	localPathIsDir := strings.HasSuffix(localPath, string(os.PathSeparator))
+	remotePathIsDir := strings.HasSuffix(remotePath, string(os.PathSeparator))
+
+	if localPathIsDir && !pull {
+		ok, err := sanityPrompt(fmt.Sprintf("The local directory '%s' will overwrite any existing contents in the remote directory '%s'", localPath, remotePath))
+		if err != nil {
+			return false, errors.Wrap(err, "error while asking user to sanity check rsync command")
+		}
+		if !ok {
+			return false, nil
+		}
+	}
+	if remotePathIsDir && pull {
+		ok, err := sanityPrompt(fmt.Sprintf("The remote directory '%s' will overwrite any existing contents in the local directory '%s'", remotePath, localPath))
+		if err != nil {
+			return false, errors.Wrap(err, "error while asking user to sanity check rsync command")
+		}
+		if !ok {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// sanityPrompt prints the message and asks the user if they would like to
+// continue or exit.
+func sanityPrompt(msg string) (ok bool, err error) {
+	fmt.Printf(msg + " Continue? (y/n) ")
+	r := bufio.NewReader(os.Stdin)
+	input, err := r.ReadString('\n')
+	if err != nil {
+		return false, errors.Wrap(err, "could not read input")
+	}
+	input = strings.TrimSpace(input)
+	return input == "y" || input == "yes", nil
+}
+
+type rsyncOpts struct {
+	local                string
+	remote               string
+	user                 string
+	host                 string
+	makeRemoteParentDirs bool
+	excludePatterns      []string
+	shouldDelete         bool
+	pull                 bool
+	dryRun               bool
+}
+
+func (opts *rsyncOpts) validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(opts.local == "", "local path must be specified")
+	catcher.NewWhen(opts.remote == "", "remote path must be specified")
+	return catcher.Resolve()
+}
+
+// buildRsyncCommand takes the given options and constructs an rsync command.
+func buildRsyncCommand(ctx context.Context, opts rsyncOpts) (*exec.Cmd, error) {
```

Comment:
should this use jasper.Command?

Reply:
I don't see a reason why jasper.Command would be superior to just using exec.CommandContext. If anything, it's more cumbersome to use because you can't do any introspection on the fields you set in the jasper.Command, which just makes testing it harder.

Reply:
just consistency and easier to program. If there's ways to improve readability (or introspect the jasper/options.Create?) that might be a good addition.